pt/lessons/advanced/metaprogramming.md

```diff
@@ -0,0 +1,246 @@
+
+Para demonstrar isso, vamos fazer um logger simples que pode ser ativado ou desativado:
+
+```elixir
+defmodule Logger do
+  defmacro log(msg) do
+    if Application.get_env(:logger, :enabled) do
+      quote do
+        IO.puts("Logged message: #{unquote(msg)}")
+      end
+    end
+  end
+end
+
+defmodule Example do
+  require Logger
+
+  def test do
+    Logger.log("This is a log message")
+  end
+end
+```
+
+Com o registo ativado a nossa fuçao `test` resultaria em um código procurando algo parecido com isto:
+
+```elixir
+def test do
+  IO.puts("Logged message: #{"This is a log message"}")
+end
+```
+
+Mas se desativarmos o log, o código resultante seria:
+
+```elixir
+def test do
+end
+```
+
+### Macros Privados
+
+Embora não seja tão comum, Elixir suporta macros privadas. Um macro privado é definido com `defmacro` e só pode ser chamado a partir do módulo no qual ele foi definido. Macros privados devem ser definidas antes do código que as invoca.
+
+### <a name="higienizacao-de-macro">Higienização de Macro
+
+A característica de como macros interagem com o contexto de quem o chamou quando expandido é conhecida como a higienização de macro. Por padrão macros no Elixir são higiênicos e não entrarão em conflito com nosso contexto:
+
+```elixir
+defmodule Example do
+  defmacro hygienic do
+    quote do: val = -1
+  end
+end
+
+iex> require Example
+nil
+iex> val = 42
+42
+iex> Example.hygienic
+-1
+iex> val
+42
+```
+
+Mas e se quisermos manipular o valor de `val` ? Para marcar uma variável como sendo anti-higiênica podemos usar `var!/2`. Vamos atualizar o nosso exemplo para incluir outro macro utilizando `var/2`!
+
+```elixir
+defmodule Example do
+  defmacro hygienic do
+    quote do: val = -1
+  end
+
+  defmacro unhygienic do
+    quote do: var!(val) = -1
+  end
+end
+```
+
+Vamos comparar a forma como eles interagem com nosso contexto:
+
+```elixir
+iex> require Example
+nil
+iex> val = 42
+42
+iex> Example.hygienic
+-1
+iex> val
+42
+iex> Example.unhygienic
+-1
+iex> val
+-1
+```
+
+Ao incluir `var!/2` em nossa macro que manipular o valor de `val` sem passá-la em nossa macro. O uso de macros não higiênicos deve ser mantido a um mínimo. Ao incluir `var!/2` que aumentam o risco de um conflito de resolução de variável.
+
+### Binding
+
+Nós já cobrimos a utilidade do `unquote/1` mas não há outra maneira de injetar valores em nosso código: *binding*. Com *binding* de variável somos capazes de incluir múltiplas variáveis em nossa macro e garantir que eles são *unquote* apenas uma vez, evitando reavaliações acidentais. Para usar variáveis de vinculação precisamos passar uma lista de palavras-chave a opção `bind_quoted` no `quote/2`.
```

Comment:
I think this `a opção` could be `à opção` or could be rewrite to `para a opção `bind_quote` de `quote/2`.